_delphi_utils_python/delphi_utils/smooth.py

```diff
@@ -0,0 +1,434 @@
+"""
+This file contains the smoothing utility functions. We have a number of
+possible smoothers to choose from: windowed average, local weighted regression,
+and a causal Savitzky-Golay filter.
+
+Code is courtesy of Dmitry Shemetov, Maria Jahja, and Addison Hu.
+
+These smoothers are all functions that take a 1D numpy array and return a smoothed
+1D numpy array of the same length (with a few np.nans in the beginning). See the
+docstrings for details.
+"""
+
+import warnings
+
+import numpy as np
+
+
+class Smoother:
+    """
+    This is the smoothing utility class. This class holds the parameter settings for its smoother
+    methods and provides reasonable defaults. Basic usage can be found in the examples below.
+
+    The smoother function takes numpy arrays as input, expecting the values to come from a
+    regularly-spaced time grid. NANs are ok, as long as the array does not begin with a NAN. The
```

Comment:
Are you expecting to see time series beginning with NaN? If so, it might be worth explicitly trimming them off rather than worrying about how they might cause problems.

Reply:
I think I have an error if the series begins with a NaN. I wanted the smooth methods to preserve the length of the array, expecting the user to insert the output back into a dataframe with a fixed index.

Reply:
Line 415. Though now I think it makes sense to pull that check out to `__init__`.